Tests - EbooksComParser

KPP13 opened the pull request:
Some tests for EbooksComParser added and refactored.

KPP13 committed:
Tests for Book, Author and PrintHouse added.
Added tests to check aquality and hashcodes.

KPP13 committed:
Fixing bugs (float ==) and code smells.
Nothing special in this commit, just some cleaning and fixing.

KPP13 committed:
Tests for EbooksComParser refactored. New tests added.
Now, three different cases are covered:
	1. the web page does not contain any books
	2. the web page does not contain books on discount
	3. the web page contains free books or books on discount
In tests samples of ebooks.com web pages are used. They are stored in test/resources directory.

KPP13 committed:
Simple refactor of the test for EbooksComParser.
Code duplication avoided.

KPP13 committed:
Tests for EbooksComParser finished.
Now, the content of the list returned by parse() method is also checked.

clinton56 commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -93,4 +96,35 @@ public String toString () {
+            return false;
+        Book book = (Book)bookObj;
+        return title.equals(book.title)
+                && abs(newPrice - book.newPrice) < 1e-10 && abs(oldPrice - book.oldPrice) < 1e-10
```
magic :1234: ? :ghost:

KPP13 replied:
No, everything is fine :) This approach is much safer because of numerical representation of float numbers -> IEEE 754.
Of course, I maybe should change epsilon to lower value, but when we consider prices it should be ok.

clinton56 commented on robot-db/src/test/java/domain/AuthorTest.java:
```diff
@@ -0,0 +1,75 @@
+import static org.testng.Assert.assertEquals;
+
+/**
+ * Tests for Author class.
```
It's obvious :tongue:

KPP13 replied:
Not for Sonar :cactus:

clinton56 replied:
Ok, I didn't know :octopus:

ghost commented:
Do you really need `hashcode` and `equals` in every class? For me there was too many comments.
Overall it looks ok. Big :+1: for tests :wink:

danone771909 commented on robot-db/src/main/java/domain/Book.java:
```diff
@@ -93,4 +96,35 @@ public String toString () {
+                && printHouse.equals(book.printHouse) && authors.equals(book.authors)
+                && category == book.category;
+    }
+}
```
Why so many && ? :smile:

KPP13 replied:
I know that I can split it into multiple lines using a helper boolean variable but I am not sure if it has a sense. 
When the false is found int the expression, next boolean values are not calculated (power of && operator :smile_cat: ). I think it is much faster than using helper variable, but I can test this later.
The drop of performance would be caused by & operator in the case when there are many false values in the expression.
What can I do is to move abs(...) to the end of the expression. JVM will not have to calculate difference, absolute value and compare when there was a false value earlier.

danone771909 replied:
I think that for now it's good. We can test it later but it works fast and maybe it's the best solution for this case. I just wanted to know why and now I know that it has sense :+1:

danone771909 commented on robot-db/src/test/java/domain/BookTest.java:
```diff
@@ -0,0 +1,128 @@
+    private final Book b2 = new Book();
+    private final Book b3 = new Book();
+
+    @BeforeClass
```
Maybe some JavaDocs would be very helpful :sos:

KPP13 replied:
Ok, I will change a description soon.

danone771909 commented on robot-logic/src/test/java/logic/parser/EbooksComParserTest.java:
```diff
@@ -21,24 +25,129 @@
+
+        // when
+        when(parser.openDocument()).thenReturn(doc);
+        when(parser.parse()).thenCallRealMethod();
```
thenCallRealMethod() it is very powerful and helpfull :+1: I really like it

KPP13 committed:
Performance refactor (boolean expression order).
Now, it should be work faster if the case when there are many false values in the expression.
Some words added to tests' javadocs - but nothing interesting - I don't have an idea now how to describe those tests.

danone771909 commented:
All are corrected :+1:
